design-docs/modes.md

```diff
@@ -0,0 +1,622 @@
+- `int GetSelectedTeleoperatedMode()` - gets the mode ID currently selected for teleoperated; valid when the robot is disabled; returns 0 if unknown or no selection
+- `int GetSelectedTestMode()` - gets the mode ID currently selected for test; valid when the robot is disabled; returns 0 if unknown or no selection
+
+## DS Protocol
+
+The list of modes (and details thereof) is communicated from the Robot to the DS via the tagged TCP link.
+
+As in the current FRC protocol, the enabled state and teleop/auto/test selection are sent as part of the UDP control word from the DS to the Robot. Additionally, the selected modes for auto, teleop, and test are sent as part of every UDP packet.  This is done via a 64-bit hash of each mode's name.  Hash collisions are extremely low probability given the low quantity of modes, but this can be checked on the robot side when generating the mode list, and spaces appended as necessary to the provided names to uniquify the hashes.
+
+# Drawbacks
+
+# Alternatives
+
+Use SendableChooser for more modes (teleop and test as well as autonomous); downsides of this:
+- Doesn't allow flexibility of mixing linear and periodic modes
+- Harder to use than annotations (generic class)
+- No grouping, no filtering by mode (although these could be added)
+
+# Trades
+
+- Overall naming: mode? routine? opmode?
+
+- Naming of mode functions? start-periodic-end, vs init-periodic-exit (2015 FRC PeriodicRobot), vs init-execute-end (2015 FRC Command), vs init-start-loop-stop (2015 FTC OpMode; note init behaves like the constructor here)
+
+- For matches, should we construct teleop at the same time as auto?  If we do that, we probably need a disabledStart() or 2015 FTC opmode style init() that's run on teleop after auto completes, and don't run disabledPeriodic for both simultaneously.
+
+- Is it actually likely that team robots will have a mix of command-based modes and non-command-based?  If not, restricting command-based robots to *only* command-based modes might enable greater integration and ease-of-use for the typical command-based robot?
+
+# Unresolved Questions
+
+- FRC SendableChooser has a "default" option set by robot code.  Do we want something similar here or should it be 100% DS driven?  It's kind of nice to be able to set a default (e.g. via a `setDefaultAutonomousMode(String)` function in `Robot`), but also might be a little fragile since it's name based. If it's done via annotation, what happens if multiple annotations are marked as default?
+
+- Should it be possible to have multiple top-level Robot classes (e.g. for different robot configurations), and have that be selectable at the DS as well?  This is a bit ugly to support even if when the Robot object is passed into the mode constructor, because different Robot class types will only work with certain modes.
```

Comment:
```suggestion
- Should it be possible to have multiple top-level Robot classes (e.g. for different robot configurations), and have that be selectable at the DS as well?  This is a bit ugly to support even if the Robot object is passed into the mode constructor, because different Robot class types will only work with certain modes.
```
I think that's how this should read?